File: etc/c#/BaseX.cs
/*
 * ----------------------------------------------------------------------------- 
 * 
 * This C# module provides methods to connect to and communicate with the
 * BaseX Server.
 *
 * The Constructor of the class expects a hostname, port, username and password
 * for the connection. The socket connection will then be established via the
 * hostname and the port.
 *
 * For the execution of commands you need to call the execute() method with the
 * database command as argument. The method returns a boolean, indicating if
 * the command was successful. The result can be requested with the result()
 * method, and the info() method returns additional processing information
 * or error output.
 *
 * -----------------------------------------------------------------------------
 * Example:
 * 
 * using System;
 * 
 * namespace BaseX
 *	{
 *	public class Example
 *	{	
 *		public static void Main(string[] args)
 *		{	
 *			try {
 *				Session session = new Session("localhost", 1984, "admin", "admin");
 *				if ((session.execute("xquery 1 to 10")) == "\0") {
 *					Console.WriteLine(s.res());
 * 				} else {
 *					Console.WriteLine(s.inf());
 *				}
 *				session.close();
 *			} catch (Exception e) {
 *				Console.WriteLine(e.Message);
 *		  }
 *		}
 *	  }
 *	} 
 * 
 * -----------------------------------------------------------------------------
 * (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License
 * -----------------------------------------------------------------------------
 */
using System;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Collections.Generic;
using System.IO;

namespace BaseX
{
	class Session
	{
		public NetworkStream stream = null;
		public TcpClient socket = null;
		public MemoryStream result = null;
		public string info = "";
		public int bpos = 0;
		public int bsize = 0;
		public byte[] inStream = new byte[4096];
		
		/** Constructor, creating a new socket connection. */
		public Session(string host, int port, string username, string pw) 
		{
			socket = new TcpClient(host, port);
			stream = socket.GetStream();
			init();
			string ts = readString();
			string h = md5(pw);
			string hts = h + ts;
			string end = md5(hts);
			send(username + "\0");
			send(end + "\0");
			if (stream.ReadByte() != 0) {
			 	throw new Exception("Access denied.");
			}
		}
		
		/** Executes the specified command. */
		public bool execute(string com, MemoryStream ms) {
			send(com + "\0");
			init();
			readString(ms);
			info = readString();
			return read() == 0;
		}
		
		/** Executes the specified command. */
		public bool execute(string com) {
			result = new MemoryStream();
			return execute(com, result);
		}
		
		/** Returns the result. */
		public string res() {
			return System.Text.Encoding.UTF8.GetString(result.GetBuffer());
		}
		
		/** Returns the processing information. */
		public string inf() {
			return info;
		}
		
		/** Closes the connection. */
		public void close() {
			send("exit \0");
			socket.Close();
		}
		
		/** Initializes the byte transfer. */
		private void init() {
			bpos = 0;
			bsize = 0;
		}
		
		/** Returns a single byte from the socket. */
		private byte read() {
			if (bpos == bsize) {
				bsize = stream.Read(inStream, 0, 4096);
				bpos = 0;
			}
			byte b = inStream[bpos];
			bpos += 1;
			return b;
		}
		
		/** Receives a string from the socket. */
		private void readString(MemoryStream ms) {
			while (true) {
				byte b = read();
				if (b != 0) {
					ms.WriteByte(b);
				} else {
					break;
				}
			}
		}
		
		/** Receives a string from the socket. */
		private string readString() {
			List<byte> list = new List<byte>();
			while (true) {
				byte b = read();
				if (b != 0) {
					list.Add(b);
				} else {
					return System.Text.Encoding.UTF8.GetString(list.ToArray());
				}
			}
		}
		
		/** Sends strings to server. */
		private void send(string message) {
			byte[] outStream = System.Text.Encoding.UTF8.GetBytes(message);
			stream.Write(outStream, 0, outStream.Length);
            stream.Flush();
		}
		
		/** Returns the md5 hash of a string. */
		private string md5(string input) {
			MD5CryptoServiceProvider MD5 = new MD5CryptoServiceProvider();
			byte[] hashValue = new byte[1];
        	byte[] bytes = Encoding.UTF8.GetBytes(input);
        	hashValue = MD5.ComputeHash(bytes);
        	StringBuilder sb = new StringBuilder();

			for (var i = 0; i <= hashValue.Length - 1; i++) {
                sb.Append(hashValue[i].ToString("x2"));
            }
            return sb.ToString();
		}
	}
}

File: etc/vb/BaseX.vb
'
' -----------------------------------------------------------------------------
'  
'  This VB.Net module provides methods to connect to and communicate with the
'  BaseX Server.
' 
'  The Constructor of the class expects a hostname, port, username and password
'  for the connection. The socket connection will then be established via the
'  hostname and the port.
' 
'  For the execution of commands you need to call the Execute() method with the
'  database command as argument. The method returns a boolean, indicating if
'  the command was successful. The result is stored in the Result property,
'  and the Info property returns additional processing information or error
'  output.
' 
'  An even faster approach is to call Execute() with the database command and
'  an output stream. The result will directly be printed and does not have to
'  be cached.
'  
'  -----------------------------------------------------------------------------
'  Example:
'  
' Imports System
'
' Namespace BaseX
'	Public Class Example
'		Public Shared Sub Main(args As String())
'			Try
'				Dim session As New Session("localhost", 1984, "admin", "admin")
'
'				' Version 1: perform command and show result or error output
'				If session.Execute("xquery 1 to 10") Then
'					Console.WriteLine(session.Result)
'				Else
'					Console.WriteLine(session.Info)
'				End If
'
'				' Version 2 (faster): send result to the specified output stream
'				Dim stream As Stream = Console.OpenStandardOutput()
'				If Not session.Execute("xquery 1 to 10", stream) Then
'					Console.WriteLine(session.Info)
'				End If
'
'				session.Close()
'			Catch e As Exception
'				Console.WriteLine(e.Message)
'			End Try
'		End Sub
'	End Class
' End Namespace
'  
'  -----------------------------------------------------------------------------
'  (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License
'  -----------------------------------------------------------------------------
' 

Imports System
Imports System.Net.Sockets
Imports System.Security.Cryptography
Imports System.Text
Imports System.Collections.Generic
Imports System.IO

Namespace BaseX
	Class BaseX
		Private m_result As New MemoryStream()
		Private cache As Byte() = New Byte(4095) {}
		Private stream As NetworkStream
		Private socket As TcpClient
		Private m_info As String = ""
		Private bpos As Integer
		Private bsize As Integer

		' Constructor, creating a new socket connection. 
		Public Sub New(host As String, port As Integer, username As String, pw As String)
			socket = New TcpClient(host, port)
			stream = socket.GetStream()
			Dim ts As String = ReadString()
			Send(username)
			Send(MD5(MD5(pw) & ts))
			If stream.ReadByte() <> 0 Then
				Throw New IOException("Access denied.")
			End If
		End Sub

		' Executes the specified command. 
		Public Function Execute(com As String, ms As Stream) As Boolean
			Send(com)
			Init()
			ReadString(ms)
			m_info = ReadString()
			Return Read() = 0
		End Function

		' Executes the specified command. 
		Public Function Execute(com As String) As Boolean
			m_result = New MemoryStream()
			Return Execute(com, m_result)
		End Function

		' Returns the result. 
		Public ReadOnly Property Result() As String
			Get
				Return System.Text.Encoding.UTF8.GetString(m_result.ToArray())
			End Get
		End Property

		' Returns the processing information. 
		Public ReadOnly Property Info() As String
			Get
				Return m_info
			End Get
		End Property

		' Closes the connection. 
		Public Sub Close()
			Send("exit")
			socket.Close()
		End Sub

		' Initializes the byte transfer. 
		Private Sub Init()
			bpos = 0
			bsize = 0
		End Sub

		' Returns a single byte from the socket. 
		Private Function Read() As Byte
			If bpos = bsize Then
				bsize = stream.Read(cache, 0, 4096)
				bpos = 0
			End If
			Return cache(bpos++)
		End Function

		' Receives a string from the socket. 
		Private Sub ReadString(ms As Stream)
			While True
				Dim b As Byte = Read()
				If b = 0 Then
					Exit While
				End If
				ms.WriteByte(b)
			End While
		End Sub

		' Receives a string from the socket. 
		Private Function ReadString() As String
			Dim ms As New MemoryStream()
			ReadString(ms)
			Return System.Text.Encoding.UTF8.GetString(ms.ToArray())
		End Function

		' Sends strings to server. 
		Private Sub Send(message As String)
			Dim msg As Byte() = System.Text.Encoding.UTF8.GetBytes(message)
			stream.Write(msg, 0, msg.Length)
			stream.WriteByte(0)
		End Sub

		' Returns the md5 hash of a string. 
		Private Function MD5(input As String) As String
			Dim MD5 As New MD5CryptoServiceProvider()
			Dim hash As Byte() = MD5.ComputeHash(Encoding.UTF8.GetBytes(input))

			Dim sb As New StringBuilder()
			For Each h As Byte In hash
				sb.Append(h.ToString("x2"))
			Next
			Return sb.ToString()
		End Function
	End Class
End Namespace
